gui/workorder-finetune

```diff
@@ -0,0 +1,544 @@
+        })
+    end
+    
+    local isActiveProduct = function(args)
+        return (args.job_item.has_material_reaction_product == args.product_materials)
+    end
+    table.insert(choices, {
+            icon = '!',
+            text = {{text = 'unset producing', pen = pen_action, dpen = pen_action_d}},
+            product_materials = ''
+        })
+    for _, product_materials in ipairs(product_materials) do
+        local args = {job_item = job_item, product_materials=product_materials}
+        table.insert(choices, {
+            icon = ' ',
+            text = {{text = product_materials .. '-producing',
+                    pen = curry(pen_cb, args, isActiveProduct),
+                    dpen = curry(pen_d_cb, args, isActiveProduct)
+            }},
+            product_materials = product_materials
+        })
+    end
+
+    set_search_keys(choices)
+    args.choices = choices
+
+    if args.on_select then
+        local cb = args.on_select
+        args.on_select = function(idx, obj)
+            return cb(obj)
+        end
+    end
+
+    return dlg.ListBox(args)
+end
+-- end gui.materials.ItemTraitDialog
+
+JobDetails = defclass(JobDetails, gui.FramedScreen)
+
+JobDetails.focus_path = 'workorder-finetune'
+
+JobDetails.ATTRS {
+    job = DEFAULT_NIL,
+    frame_inset = 1,
+    frame_background = COLOR_BLACK,
+}
+
+function JobDetails:init(args)
+    self:addviews{
+        widgets.Label{
+            frame = { l = 0, t = 0 },
+            text = {
+                { text = df.job_type.attrs[self.job.job_type].caption }, NEWLINE, NEWLINE,
+                '  ', status
+            }
+        },
+        widgets.Label{
+            frame = { l = 0, t = 4 },
+            text = {
+                { key = 'CUSTOM_I', text = ': Input item, ',
+                  enabled = self:callback('canChangeIType'),
+                  on_activate = self:callback('onChangeIType') },
+                { key = 'CUSTOM_M', text = ': Material, ',
+                  enabled = self:callback('canChangeMat'),
+                  on_activate = self:callback('onChangeMat') },
+                { key = 'CUSTOM_T', text = ': Traits',
+                  enabled = self:callback('canChangeTrait'),
+                  on_activate = self:callback('onChangeTrait') }
+            }
+        },
+        widgets.List{
+            view_id = 'list',
+            frame = { t = 6, b = 2 },
+            row_height = 4,
+        },
+        widgets.Label{
+            frame = { l = 0, b = 0 },
+            text = {
+                { key = 'LEAVESCREEN', text = ': Back',
+                  on_activate = self:callback('dismiss') }
+            }
+        },
+    }
+
+    self:initListChoices()
+end
+
+function JobDetails:onGetSelectedJob()
+    return self.job
+end
+
+describe_item_type = wsj.describe_item_type
+is_caste_mat = wsj.is_caste_mat
+describe_material = wsj.describe_material
+list_flags = wsj.list_flags
+
+function describe_item_traits(iobj)
```

Comment:
consider declaring these as `local` so they don't pollute the exported environment (also `set_search_keys` above)

Reply:
did that too